onnxruntime/core/mlas/lib/sqnbitgemm_kernel_avx512.cpp

```diff
@@ -0,0 +1,602 @@
+    }
+    blob += BlkLen;
+  }
+}
+
+void MLASCALL
+QuantizeARow_CompInt8_avx512(
+  size_t BlkLen,
+  const float* A,
+  size_t CountK,
+  std::byte* QuantA
+)
+{
+  MlasQ80BlkQuantRow_avx512(BlkLen, A, CountK, QuantA);
+}
+
+MLAS_FORCEINLINE
+void
+SQ4BitGemmM1Kernel_CompInt8_avx512vnni(
+  size_t BlkLen,
+  const std::byte* QuantA,
+  const std::byte* QuantBData,
+  const float* QuantBScale,
+  const std::byte* QuantBZeroPoint,
+  float* C,
+  size_t CountN,
+  size_t CountK,
+  size_t BlockStrideQuantB,
+  const float* Bias
+)
+{
+  if (QuantBZeroPoint != nullptr) {
+    constexpr bool HasZeroPoint = true;
+    if (BlkLen == 16) {
+      SQ4BitGemmM1Kernel_BlkLen16_CompInt8_Impl<HasZeroPoint>(
+        QuantA,
+        QuantBData,
+        QuantBScale,
+        QuantBZeroPoint,
+        C,
+        CountN,
+        CountK,
+        BlockStrideQuantB,
+        Bias
+      );
+    }
+    else if (BlkLen == 32) {
+      SQ4BitGemmM1Kernel_BlkLen32_CompInt8_Impl<HasZeroPoint, accumulate_mul_sum_avx512vnni<HasZeroPoint>>(
+        QuantA,
+        QuantBData,
+        QuantBScale,
+        QuantBZeroPoint,
+        C,
+        CountN,
+        BlockStrideQuantB,
+        Bias
+      );
+    }
+    else {
+      SQ4BitGemmM1Kernel_BlkLen64Plus_CompInt8_Impl<HasZeroPoint, dot_quad_avx512vnni>(
+        BlkLen,
+        QuantA,
+        QuantBData,
+        QuantBScale,
+        QuantBZeroPoint,
+        C,
+        CountN,
+        CountK,
+        BlockStrideQuantB,
+        Bias
+      );
+    }
+  }
+  else {
+    constexpr bool HasZeroPoint = false;
+    if (BlkLen == 16) {
+      SQ4BitGemmM1Kernel_BlkLen16_CompInt8_Impl<HasZeroPoint>(
+        QuantA,
+        QuantBData,
+        QuantBScale,
+        QuantBZeroPoint,
+        C,
+        CountN,
+        CountK,
+        BlockStrideQuantB,
+        Bias
+      );
+    }
+    else if (BlkLen == 32) {
+      SQ4BitGemmM1Kernel_BlkLen32_CompInt8_Impl<HasZeroPoint, accumulate_mul_sum_avx512vnni<HasZeroPoint>>(
+        QuantA,
+        QuantBData,
+        QuantBScale,
+        QuantBZeroPoint,
+        C,
+        CountN,
+        BlockStrideQuantB,
+        Bias
+      );
+    }
+    else {
+      SQ4BitGemmM1Kernel_BlkLen64Plus_CompInt8_Impl<HasZeroPoint, dot_quad_avx512vnni>(
+        BlkLen,
+        QuantA,
+        QuantBData,
+        QuantBScale,
+        QuantBZeroPoint,
+        C,
+        CountN,
+        CountK,
+        BlockStrideQuantB,
+        Bias
+      );
+    }
+  }
+}
+
+const MLAS_SQNBIT_GEMM_DISPATCH MlasSQNBitGemmDispatchAvx512 = []() {
+    MLAS_SQNBIT_GEMM_DISPATCH d;
+
+    d.SQ4BitGemmPackQuantBDataSize = SQ4BitGemmPackQuantBDataSize;
+    d.SQ4BitGemmPackQuantBData = SQ4BitGemmPackQuantBData;
+
+    d.SQ4BitGemmM1Kernel_CompFp32 = SQ4BitGemmM1Kernel_CompFp32_avx512;
+    d.Q4BitBlkDequantBForSgemm_CompFp32 = Q4BitBlkDequantBForSgemm_CompFp32_avx2;
+
+    d.SQ4BitGemmM1Kernel_CompInt8 = SQ4BitGemmM1Kernel_CompInt8_avx512vnni;
```

Comment:
>SQ4BitGemmM1Kernel_CompInt8_avx512vnni [](http://example.com/codeflow?start=36&length=38)

the separation of avx512 and avx512vnni are not implemented?

Reply:
moved to vnni.